doc/nat/nat_design_spec.md

```diff
@@ -0,0 +1,1517 @@
+- The iptables and connection tracking modules in the kernel are leveraged to handle the NAT miss packets to create the NAT/NAPT entries, that are added to the hardware.
+
+## 3.2 DB Changes
+### 3.2.1 Config DB
+Following table changes are done in Config DB.
+
+#### 3.2.1.1 STATIC_NAPT Table
+```
+STATIC_NAPT|{{global_ip}}|{{ip_protocol}}|{{global_l4_port}} 
+    "local_ip": {{ip_address}}
+    "local_port" :{{l4_port}}
+    "nat_type"   :{{snat-or-dnat}}      (OPTIONAL)
+    "twice_nat_id" :{{twice-nat-id}}    (OPTIONAL)
+```
+
+```
+This config tells to do:
+- If the "nat_type" is 'dnat':
+  - DNAT translation of the DIP/DPORT in the IP packet from 'global_ip' address and 'global_port' to 'local_ip' address and 'local_l4_port' when the packet crosses the zones.
+  - SNAT translation of the SIP/SPORT in the IP packet from 'local_ip' address and 'local_port'
+  to 'global_ip' address and 'global_l4_port' when the packet crosses the zones.
+- If the "nat_type" is 'snat':
+  - SNAT translation of the SIP/SPORT in the IP packet from 'global_ip' address and 'global_port' to 'local_ip' address and 'local_l4_port' when the packet crosses the zones.
+  - DNAT translation of the DIP/DPORT in the IP packet from 'local_ip' address and 'local_port'
+  to 'global_ip' address and 'global_l4_port' when the packet crosses the zones.
+- The default value of nat_type is 'dnat' if the option is not given.
+```
+
+#### 3.2.1.2 STATIC_NAT Table
+```
+STATIC_NAT|{{global_ip}} 
+    "local_ip": {{ip_address}}
+    "nat_type"   :{{snat-or-dnat}}      (OPTIONAL)
+    "twice_nat_id" :{{twice-nat-id}}    (OPTIONAL)
+```
+
+```
+This config tells to do:
+- If the "nat_type" is 'dnat':
+  - DNAT translation of the DIP in the IP packet from 'global_ip' address to 'local_ip' address when the packet crosses the zones.
+  - SNAT translation of the SIP in the IP packet from 'local_ip' address to 'global_ip' address when the packet crosses the zones.
+- If the "nat_type" is 'snat':
+  - SNAT translation of the SIP in the IP packet from 'global_ip' address to 'local_ip' address when the packet crosses the zones.
+  - DNAT translation of the DIP in the IP packet from 'local_ip' address to 'global_ip' address when the packet crosses the zones.
+- The default value of nat_type is 'dnat' if the option is not given.
+```
+#### 3.2.1.3 NAT_GLOBAL Table
+```
+NAT_GLOBAL|Values
+    "admin_mode"     : {{enable-or-disable}}
+    "nat_timeout"    : {{timeout_in_secs}}
+    "nat_tcp_timeout": {{timeout_in_secs}}
+    "nat_udp_timeout": {{timeout_in_secs}}
+```
+
+#### 3.2.1.4 NAT_POOL Table
+```
+NAT_POOL|{{pool_name}} 
+    "nat_ip": {{ip_address}} (or) {{ip_addr_start}}-{{ip_addr_end}} 
+    "nat_port": {{start_l4_port}}-{{end_l4_port}}  (OPTIONAL)
+```
+
+#### 3.2.1.5 ACL to Pool Bindings
+```
+NAT_BINDINGS|{{binding-name}}
+    "nat_pool": {{pool-name}}
+    "access_list": {{access-list-name}} (OPTIONAL) 
+    "nat_type": {{snat-or-dnat}}		(OPTIONAL)
```

Comment:
NAT bindings are used for dynamic NAT/NAPT. Having the NAT type as SNAT means, the Source addresses are dynamically translated when the bindings are applied. We currently support only SNAT type for dynamic pool bindings.
In future, we also can add support for dynamically doing DNAT (use case being FQDN service with session distribution at the firewall https://docs.paloaltonetworks.com/pan-os/8-1/pan-os-admin/networking/nat/source-nat-and-destination-nat/destination-nat.html)..

Reply:
In the ACL there is only a do-not-nat action proposed. So how is the do-not-nat going to distinguish between nat_type for which it should disable translation? E. g. for this configuration:
```
ACL_TABLE|10
    "stage": "INGRESS", 
    "type": "L3", 
    "policy_desc": "nat-acl", 
    "ports": "Vlan2000"

ACL_RULE|10|1
    "priority": "20", 
    "src_ip": "20.0.1.0/24",
    "packet_action": "do_not_nat"

ACL_RULE|10|2
    "priority": "10", 
    "src_ip": "20.0.0.0/16", 
    "packet_action": "forward"

NAT_BINDINGS|nat1    
    "access_list": "10"
    "nat_pool": "pool1
    "nat_type": "snat"
```
How do we tell HW to disable only SNAT for packets with SRC_IP "20.0.1.0/24" and to not disable DNAT for the same packet?

Reply:
ACL action do-not-nat disables any NAT that can happen on the packet.
And the nat-type attribute in the bindings config is only of use for the kernel iptables to act upon (whether to apply or not apply snat or dnat) on the NAT miss traffic to the CPU.
And like mentioned above, only snat type is supported in the bindings.

Reply:
So in the example above, if a packet with the SRC IP from 20.0.1.0/24 and DST IP from the DNAT_POOL, won't be translated and/or trapped to CPU because of that ACL with do-not-nat action.
Is it desired behavior? If the application controls SNAT and DNAT separately, why not do the same in HW?

Reply:
@marian-pritsak , in NAT SAI pipeline, SNAT and DNAT are also controlled separately. Please raise again if this should be discussed further.

Reply:
@rlhui the DO_NOT_NAT is global in the SAI pipeline. Once you disabled the NAT, it is disabled for all the types - SNAT< DNAT, DOUBLE_NAT. I don't get why we should do it differently in SONiC.

Reply:
@marian-pritsak , you're right that if a pkt matches DO_NOT_NAT condition in the SAI pipeline, then DNAT or DOUBLE_NAT is bypassed too.  In SONiC, currently it seems acl binding including DO_NOT_NAT is supported with SNAT Only.

@kirankella, can you please confirm?  Is there a way to bypass NAT for some pkts e.g. some known protocols for static DNAT?

Reply:
@marian-pritsak @rlhui What you said is right. In SONiC, currently acl binding including DO_NOT_NAT is supported with SNAT only (ideally other vendors config apply ACLs to the SNAT dynamic pool allocation). So on those lines the iptables added apply the do-not-nat checks to exclude the some hosts during SNAT. Also, the ACL (with do-not-nat) is typically applied on all the inside zone ports where the outbound traffic is being SNAT'ted. That way do-not-nat is not applied on the outside zone ports in the hardware.

But as per the SAI pipeline, the DO_NOT_NAT condition is like a global condition i see, but that be can be achieved only if this ACL rule is applied on all the ports (both the inside zone and the outside zone ports).
Currently for the static DNAT (you are not referring to NAPT right?), all the traffic is DNAT'ted in the hardware for all protocols by the single DNAT entry. To bypass DNAT for some packets, we have to use the ACL DO_NOT_NAT rule and apply the ACL on those outside zone ports. So that they will be regularly forwarded in the hardware.

Reply:
@kirankella, thanks.  For each dynamic SNATP entry going outbound, there's at least one reverse flow, that needs the DNATP back to the source.  If DO_NOT_NAT was applied to the outbound flow, same would apply to the corresponding inbound flow.  Can this be supported?
For DO_NOT_NAT for static DNAT or DNATP, it could be a future item.

Reply:
OK